README.md
IF THIS FILE HAS NO LINE BREAKS:  View it in a web browser. 

LoROM template
==============

This is a minimal working program for the Super Nintendo
Entertainment System using the LoROM (mode $20) mapper.

Concepts illustrated:

* internal header (with correct checksum) and init code
* setting up a static background
* loading data from multiple 32 KiB banks in a LoROM
* structure of a game loop
* automatic controller reading
* 8.8 fixed-point arithmetic
* acceleration-based character movement physics
* sprite drawing and animation, with horizontal flipping
* makefile-controlled conversion of graphics to tile data in
  both 2-bit-per-pixel and 4-bit-per-pixel formats
* booting the S-SMP (SPC700 audio CPU)
* writing SPC700 code using 65C02 syntax using blargg's
  [SPC700 macro pack] for ca65
* use of SPC700 timers to control playback
* reading a sequence of pitches and converting them to frequencies
* makefile-controlled compression of sampled sound to BRR format
* creating an SPC700 state file for SPC music players

Concepts not illustrated:

* a 512-byte header for obsolete floppy-based copiers
* S-CPU/S-SMP communication to play sound effects or change songs

[SPC700 macro pack]: http://forums.nesdev.com/viewtopic.php?p=121690#p121690

Setting up the build environment
--------------------------------
Building this demo requires cc65, Python, Pillow, GNU Make, and GNU
Coreutils.  For detailed instructions to set up a build environment,
see [nrom-template].

[nrom-template]: https://github.com/pinobatch/nrom-template

Organization of the program
---------------------------

### Include files

* `snes.inc`: Register definitions and useful macros for the S-CPU
* `global.inc`: S-CPU global variable and function declarations
* `spc-ca65.inc`: Macro library to produce SPC700 instructions
* `spc-65c02.inc`: Macro library to use 65C02 syntax with the SPC700

### Source code files (65816)

* `snesheader.s`: Super NES internal header
* `init.s`: PPU and CPU I/O initialization code
* `main.s`: Main program
* `bg.s`: Background graphics setup
* `player.s`: Player sprite graphics setup and movement
* `ppuclear.s`: Useful subroutines for interacting with the S-PPU
* `blarggapu.s`: Send a sound driver to the S-SMP 

### Source code files (SPC700)

* `spcheader.s`: Header for the `.spc` file; unused in `.sfc`
* `spcimage.s`: Sound driver

Each source code file is made up of subroutines that start with
`.proc` and end with `.endproc`.  See the [ca65 Users Guide] for
what these mean.

[ca65 Users Guide]: http://cc65.github.io/doc/ca65.html

The tools
---------
The `tools` folder contains a few essential command-line programs
written in Python to convert asset data (graphics, audio, etc.) into
a form usable by the Super NES.  The makefile contains instructions
to run these programs whenever the original asset data changes.

* `pilbmp2nes.py` converts bitmap images in PNG or BMP format
  into tile data usable by several classic video game consoles.
  It has several options to control the data format; use
  `pilbmp2nes.py --help` from the command prompt to see them all.
* `wav2brr.py` converts an uncompressed wave file to the BRR (bit
  rate reduction) format, a lossy audio codec based on ADPCM used
  by the S-DSP (the audio chip in the Super NES).
* `karplus.py` generates a plucked string sound, used for the
  bass sample.
* `hihat.py` generates a noise sample.

Greets
------

* [Super Nintendo Development Wiki] contributors
* Martin Korth (nocash) for [Fullsnes] doc and [NO$SNS] emulator
* Shay Green (blargg) for APU examples and SPC700 macro pack
* Jeremy Chadwick (koitsu) for more code organization tips

[Super Nintendo Development Wiki]: http://wiki.superfamicom.org/
[Fullsnes]: http://problemkaputt.de/fullsnes.htm
[NO$SNS]: http://problemkaputt.de/sns.htm

Legal
-----
The demo is distributed under the zlib License, reproduced below:

> Copyright 2017 Damian Yerrick
> 
> This software is provided 'as-is', without any express or implied
> warranty.  In no event will the authors be held liable for any
> damages arising from the use of this software.
> 
> Permission is granted to anyone to use this software for any
> purpose, including commercial applications, and to alter it and
> redistribute it freely, subject to the following restrictions:
> 
> 1. The origin of this software must not be misrepresented; you must
>    not claim that you wrote the original software. If you use this
>    software in a product, an acknowledgment in the product
>    documentation would be appreciated but is not required.
> 
> 2. Altered source versions must be plainly marked as such, and must
>    not be misrepresented as being the original software.
> 
> 3. This notice may not be removed or altered from any source
>    distribution.

A work's "source" form is the preferred form of a work for making
modifications to it.

CHANGES.txt
0.06 (2016-11-17)
* wav2brr: option to produce a looped BRR
* wav2brr: option to skip pre- and de-emphasis filtering
* spc-ca65.inc: instructions that use DEFAULT_ABS work in
  .proc blocks
* Added Karplus-Strong synthesizer to tools
* Sets DSP registers based on note number and instrument number,
  which could form the basis for a music engine
* Converted all Python tools to Python 3
* Extra explanation for OBSEL, nmi_handler, pseudo hi-res,
  and interlace (requested by nicklausw)
* snes.inc has doc comments for seta/setxy (requested by Pokun)
* Experimental checksum correction tool, not enabled by default
* Now on GitHub

0.05 (2015-02-13): The "Ho-ly-$#!+" Code Review
* fixed branch bug in collision introduced during translation
  from 6502 to "idiomatic" 65816 (reported by thefox)
* added overall map of source code to README.md
  (requested by Espozo)
* comment readability improvements
* added doc comments before more functions (requested by Espozo)
* more consistent use of seta/setxy macros
  to change processor word size
* split code related to the background and the player
  into separate files (requested by Espozo)
* compensated for missing background line 0
* explained unfamiliar register names
* renamed .h files to .inc so as not to confuse with C headers
  (requested by koitsu)
* rewrote init code based on koitsu's InitializeSNES macro
* indented forward branches (requested by Espozo)
* moved ROM segments to $808000 to allow use of fast ROM
  (requested by koitsu)
* imported the direct page base address as a symbol
  (requested by koitsu)
* broke snes.inc into functional groups
* added register name aliases from Martin Korth's fullsnes.htm
  (requested by koitsu)

0.04a (2015-01-28)
* corrected missing lorom256k.cfg file

0.04 (2014-11-09):
* expanded to 256 KiB to demonstrate far operation
* moved parts into separate banks to show use of multiple banks
* snesheader: added symbolic names for mappers, regions, and
  memory sizes and speeds
* wav2brr.py: added proper command line options including
  decompression of BRR back to WAV
* spcimage.s: fixed stya macro (reported by doppel)

0.03 (2014-10-03):
* wav2brr.py: made NumPy dependency optional
* added other explanatory comments to main.s
* Corrected some errors related to BSS in linker config file
* snes.h: macros for controlling accumulator and index width
  (requested by Bisqwit)
* snes.h: macro for nametable X, Y location (requested by Bisqwit)
* New template for SPC files

0.02 (2014-09-24): 
* converted README to Markdown
* added instructions for installing build environment to README
  (requested by whicker)
* corrected a few factual errors in README related to differences
  between the NES and the Super NES
* added a wav file and wav to BRR converter
* removed Encyclopedia Dramatica "goes where" meme from index.txt

0.01 (2014-09-21): 
* initial release


src/main.s
.include "snes.inc"
.include "global.inc"
.smart
.export main, nmi_handler

USE_PSEUDOHIRES = 0
USE_INTERLACE = 0
USE_AUDIO = 1

.segment "ZEROPAGE"
nmis: .res 1
oam_used: .res 2

.segment "BSS"

OAM:   .res 512
OAMHI: .res 512
; OAMHI contains bit 8 of X (the horizontal position) and the size
; bit for each sprite.  It's a bit wasteful of memory, as the
; 512-byte OAMHI needs to be packed by software into 32 bytes before
; being sent to the PPU, but it makes sprite drawing code much
; simpler.  The OBC1 coprocessor used in the game Metal Combat:
; Falcon's Revenge performs the same packing function in hardware,
; possibly as a copy protection method.

.segment "CODE"
;;
; Minimalist NMI handler that only acknowledges NMI and signals
; to the main thread that NMI has occurred.
.proc nmi_handler
  ; Because the INC and BIT instructions can't use 24-bit (f:)
  ; addresses, set the data bank to one that can access low RAM
  ; ($0000-$1FFF) and the PPU ($2100-$213F) with a 16-bit address.
  ; Only banks $00-$3F and $80-$BF can do this, not $40-$7D or
  ; $C0-$FF.  ($7E can access low RAM but not the PPU.)  But in a
  ; LoROM program no larger than 16 Mbit, the CODE segment is in a
  ; bank that can, so copy the data bank to the program bank.
  phb
  phk
  plb

  seta8
  inc a:nmis       ; Increase NMI count to notify main thread
  bit a:NMISTATUS  ; Acknowledge NMI

  ; And restore the previous data bank value.
  plb
  rti
.endproc

;;
; This program doesn't use IRQs either.
.proc irq_handler
  rti
.endproc

.segment "CODE1"
; init.s sends us here
.proc main

  ; In the same way that the CPU of the Commodore 64 computer can
  ; interact with a floppy disk only through the CPU in the 1541 disk
  ; drive, the main CPU of the Super NES can interact with the audio
  ; hardware only through the sound CPU.  When the system turns on,
  ; the sound CPU is running the IPL (initial program load), which is
  ; designed to receive data from the main CPU through communication
  ; ports at $2140-$2143.  Load a program and start it running.
  .if ::USE_AUDIO
    jsl spc_boot_apu
  .endif

  jsl load_bg_tiles  ; fill pattern table
  jsl draw_bg        ; fill nametable
  jsl load_player_tiles

  ; In LoROM no larger than 16 Mbit, all program banks can reach
  ; the system area (low RAM, PPU ports, and DMA ports).
  ; This isn't true of larger LoROM or of HiROM (without tricks).
  phk
  plb

  ; Program the PPU for the display mode
  seta8
  stz BGMODE     ; mode 0 (four 2-bit BGs) with 8x8 tiles
  stz BGCHRADDR  ; bg planes 0-1 CHR at $0000

  ; OBSEL needs the start of the sprite pattern table in $2000-word
  ; units.  In other words, bits 14 and 13 of the address go in bits
  ; 1 and 0 of OBSEL.
  lda #$4000 >> 13
  sta OBSEL      ; sprite CHR at $4000, sprites are 8x8 and 16x16
  lda #>$6000
  sta NTADDR+0   ; plane 0 nametable at $6000
  sta NTADDR+1   ; plane 1 nametable also at $6000
  ; set up plane 0's scroll
  stz BGSCROLLX+0
  stz BGSCROLLX+0
  lda #$FF
  sta BGSCROLLY+0  ; The PPU displays lines 1-224, so set scroll to
  sta BGSCROLLY+0  ; $FF so that the first displayed line is line 0

  ; Pseudo hi-res and interlace are optional hardware features.
  ; This demo doesn't use them much, but you can enable them when
  ; building it to see what they look like.  They're described at
  ; http://wiki.superfamicom.org/snes/show/Registers

.if ::USE_PSEUDOHIRES
  ; set up plane 1's scroll, offset by 4 pixels, to show
  ; the half-pixels of pseudohires
  lda #4
  sta BGSCROLLX+2
  stz BGSCROLLX+2
  lda #$FF
  sta BGSCROLLY+2
  sta BGSCROLLY+2
  lda #%00000010   ; enable plane 1 for left halves
  sta BLENDSUB
phbit = SUB_HIRES  ; split horizontal pixels
.else
phbit = 0
.endif

.if ::USE_INTERLACE
ilbit = INTERLACE
.else
ilbit = 0
.endif

  lda #phbit|ilbit
  sta PPURES
  lda #%00010001  ; enable sprites and plane 0
  sta BLENDMAIN
  lda #VBLANK_NMI|AUTOREAD  ; but disable htime/vtime IRQ
  sta PPUNMI

  ; Set up game variables, as if it were the start of a new level.
  stz player_facing
  stz player_dxlo
  lda #184
  sta player_yhi
  setaxy16
  stz player_frame_sub
  lda #48 << 8
  sta player_xlo

forever:

  jsl move_player

  ; Draw the player to a display list in main memory
  setaxy16
  stz oam_used
  jsl draw_player_sprite

  ; Mark remaining sprites as offscreen, then convert sprite size
  ; data from the convenient-to-manipulate format described by
  ; psycopathicteen to the packed format that the PPU actually uses.
  ldx oam_used
  jsl ppu_clear_oam
  jsl ppu_pack_oamhi

  ; Backgrounds and OAM can be modified only during vertical blanking.
  ; Wait for vertical blanking and copy prepared data to OAM.
  jsl ppu_vsync
  jsl ppu_copy_oam
  seta8
  lda #$0F
  sta PPUBRIGHT  ; turn on rendering

  ; wait for control reading to finish
  lda #$01
padwait:
  bit VBLSTATUS
  bne padwait
  stz BGSCROLLX
  stz BGSCROLLX

  jmp forever
.endproc



